Convert to TypeScript

yuri opened the pull request:
DO NOT MERGE

This is not ready for merging yet, but I was wondering if you guys have any thoughts on this so far. The goal here is to get closer to Angular 2 style while using Angular 1.

The one big thing that is missing at this point is that unit tests have not been updated. Also, once the code is ready, all the handouts would need to be updated.

@e-schultz @SethDavenport @winkerVsBecks

yuri commented:
@brianolynyk This is sort of what I was thinking for the next week's project too.

SethDavenport commented on client/app/app.js:
```diff
@@ -7,7 +13,16 @@ angular.module('ngcourse', [
-])
+]);
+
+angular.module('ngcourse.tasks', ['koast']).service('tasks', TasksService);
```
I like this pattern - treating app.js as a DI wiring file.

brianolynyk replied:
I second this pattern.  It will make our file structure simpler too since all module creation will happen in this file instead of being spread across multiple files.  Also, traceability for new people coming into the app.  They can see the all the definitions and easily see what files they come from.

e-schultz replied:
Wondering how well this would scale with a larger application. Shape is getting pretty large - and by having every controller/service/etc wired up like this in here would quickly get unwieldy.

SethDavenport replied:
We could scale it by splitting it into angular sub-modules, and having a wiring file for each sub-module.

SethDavenport commented on client/app/core/tasks/tasks-service.js:
```diff
@@ -1,39 +1,42 @@
+    );
+  }
+}
+TasksService.$inject = ['koast'];
```
At one point I played around with ES6 and angular and was able to get injection working using a bit of hackery - injectables would be constructor arguments and then in my app JS I did something like

```javascript
angular.module('app').service((koast) -> { return new TasksService(koast); });
```

The arrow function is needed to get Angular's DI regex stuff to work.  However it has the advantage of keeping all DI knowledge in the wiring file (app.js) instead of having some of it in app.js and some of it in each JS module.

Removing the need for `$inject` in the module file is nice too because it means that the service implementation is closer to 'pure' JavaScript: fewer explicit angular 1.0 dependencies scattered all over the code base.  This makes migration to another framework (Angular 2.0) easier.

SethDavenport replied:
See here for an example: https://github.com/SethDavenport/visualizations/blob/es6/scripts/main.js

yuri replied:
@SethDavenport I don't understand how your solution works.

```js
.factory('Circle', function(Point) { return Circle; })
```

How does Circle end up getting access to Point? It looks like it would just get dropped.

yuri replied:
Also, did you have a chance to look at how I am handling this with InjectableReceiver?

SethDavenport replied:
This is a .factory, so it's expected to return a constructor.  In ES6 a reference to the class `Circle` returns it's constructor, which takes point as a parameter (not seen here).

However in order to get Angular 1's DI to inject Point when it actually calls the constructor, I wrap it in a function with an explicit arg list.

A better approach might have been:

```javascript
.service('Circle', function(Point) { return new Circle(Point); })
```

This does the same thing, but it's clearer.

SethDavenport replied:
Regarding injectableReceiver I guess we're trying to solve different problems.  I'm trying to decouple angular-specific DI from the actual class implementations, whereas you're trying to insulate against parameter ordering mismatch problems?

Maybe a variation of the above but with named params would work:

```javascript
.service(function (koast, foo) { 
  return new TasksService( { 
    koast: koast,
    foo: foo
  }); 
});
```

yuri replied:
@SethDavenport This won't work because your TaskService is now longer using the Angular2 pattern. TaskService needs to take injected dependencies as separate arguments. My thinking is that using TaskService.$inject is the closest we can get to doing it the way Angular2 does without leveraging typing or decorators.

yuri replied:
@SethDavenport If we move the .$inject part up, then the difference between our Angular 1 service and Angular 2 is

```js
TasksService.$inject = ['koast'];
class TasksService extends InjectableReceiver {
  constructor () {
```

vs

```js
class TasksService {
  constructor (koast: KoastService) {
```

SethDavenport replied:
Yeah OK.  I like having the $inject = stuff at the top then.

yuri commented on client/app/core/tasks/tasks-service.js:
```diff
@@ -1,39 +1,42 @@
     return function () {
       var myArgs = arguments;
-      return koast.user.whenAuthenticated()
+      return this.services.koast.user.whenAuthenticated()
```
@SethDavenport @e-schultz @brianolynyk How do you feel about this approach of storing injected services in "this.services"? Seems a little verbose to me, but I haven't thought of a better strategy yet that would be Ang2-compatible.

yuri replied:
@SethDavenport @e-schultz @brianolynyk Some examples I've seen just add injected services to "this", but I find this a little confusing. Another option is to put them under "this.$".

e-schultz replied:
I'll take a look at this later, just saw this in my inbox now - really curious to spend a bit more time with it and see whats being done.

Saw some good examples at ng-vegas also.

e-schultz commented on client/app/sections/task-add/task-add-controller.js:
```diff
@@ -1,17 +1,20 @@
-      .then(null, $log.error);
+  save(task) {
+    return this.services.tasks.addTask(task)
+      .then(() => this.services.router.goToTaskList())
```
I understand the problem that is trying to be solved here, and the idea of InjectableReciever is good, but something bugs me about the verbosity required by this.services.serviceName.

I can see if there are any other alternative options available for handling this

yuri replied:
@e-schultz It will need to be something hanging off "this", since there are no closures with the "class" syntax. So, the options are: "this.router", "this.services.router", or "this.$.router". (Or pick something else instead of "$".)

e-schultz commented:
@SethDavenport @yuri @brianolynyk anyone take a look at https://github.com/smaye81/angular-es6-seed

yuri commented:
@e-schultz In regards to angular-es6-seed, it's taking a similar approach, but hanging injected services directly onto "this" and doing so manually. And capitalizing instances, which I wouldn't endorse.

yuri commented:
I switched over to TypeScript and implemented DI using a decorator. So, it now looks as follows:

```js
constructor(
    @Inject('$log') $log,
    @Inject('users') users,
    @Inject('koast') koast
  ) {
...
}
```

This is not the most compact way of expressing it, but it aligns fairly closely with Angular2 (except that here we are using strings as annotation values).

I also got rid of the "InjectableReceivable" class and switched to a function instead:

```js
this.services = getServices(this.constructor, arguments);
```

I am thinking this would be easier to understand: `getServices` takes a function / class and the list of injected arguments and returns those arguments as an object, assigning them to properties based on what's specified in this function's $inject.

yuri commented:
Switched some of the controllers to component-style directives. It might be worth doing the same with some of the state controllers, though then we'll need something else to illustrate states.

yuri commented:
I make things simpler, I moved this to a new repo: https://github.com/rangle/ngcourse-next/. So, I will close this PR.
